### DISCUSSION TITLE
How to access application state from sqlite user-defined function?

### DISCUSSION THREAD

**Discussion by Adam-Langley** [simolus3/drift#2714] [Q&A] [answered]
I need to access some application state from within a sqlite trigger.
I have written a custom function, which just returns a boolean flag that is managed by the application - this works, and is invoked correctly when required.
However, modifying the boolean flag state never propagates to the user defined function.

I am using NativeDatabase.createInBackground, and this is where I register my function.
I expect that the trigger is calling the function in the drift background isolate, however I am modifying the boolean flag in the main application isolate - hence they are not the same variable.

How do I cross this boundary?
I believe I need to (somehow) pass a custom message across the isolate, then receive it and parse out the updated value of the flag, so it can be returned by the user defined function lambda....
Not sure how to access everything I need.

Thank you!

**Comment by simolus3** [comment, marked as answer]
> How do I cross this boundary?
I believe I need to (somehow) pass a custom message across the isolate, then receive it and parse out the updated value of the flag, so it can be returned by the user defined function lambda....

Pretty much exactly this. Typically, the handshake is done by the the main isolate establishing a `ReceivePort` onto which the background isolate will send its own `ReceivePort` to. So you could have a wrapper class like this:

```dart
typedef StateForApplication = bool;

class StateManager {
  static StateManager instance = StateManager(false);

  SendPort? _port;
  StateForApplication _state;

  StateManager(this._state);

  SendPort listenForUpdates() {
    final receive = ReceivePort();
    receive.listen((message) {
      _state = message as StateForApplication;
    });

    return receive.sendPort;
  }

  void sendTo(SendPort port) {
    _port = port;
    port.send(_state);
  }

  StateForApplication get state => _state;
  set state(StateForApplication state) {
    _state = state;
    _port?.send(state);
  }
}
```

And then integrate it to your database like this:

```dart
  final receivePort = ReceivePort();
  receivePort.first
      .then((r) => StateManager.instance.sendTo(r as SendPort))
      .whenComplete(receivePort.close);
  final send = receivePort.sendPort;

  final db = Database(
    NativeDatabase.createInBackground(
      File('/tmp/foo.db'),
      setup: (database) {
        send.send(StateManager.instance.listenForUpdates());

        database.createFunction(
            functionName: 'state',
            function: (args) => StateManager.instance.state);
      },
    ),
  );
```

Note that this doesn't do lifecycle management. Since the background isolate creates a `ReceivePort`, it will stay active even if you decide to close the database. So if you need to close the database before the application terminates, you'd have to complicate the protocol to also encode "done" events instructing the `StateManager` on the background isolate to close its `ReceivePort`.

**Reply by Adam-Langley** [reply]
Thank you @simolus3 this works perfectly!

**Reply by Adam-Langley** [reply]
I also realised (after a sleep 😄 ) that I could have created a user function to set the state also, and use SQL to invoke that function. That would have been simpler code, however I have chosen to use yours as it gives me greater flexibility in future.

Thank you again.
